File: src/test/java/com/algo/linkedlist/ReverseListTest.java
package com.algo.linkedlist;

import org.testng.annotations.Test;

import static org.hamcrest.CoreMatchers.*;
import static org.hamcrest.MatcherAssert.assertThat;

public class ReverseListTest {

    private final ReverseList reverseList = new ReverseList();

    @Test
    public void shouldReverseOneNodeList() {
        //given
        ListNode<Integer> head = new ListNode(1);

        //when
        ListNode<Integer> reverse = reverseList.reverse(head);

        //then
        assertThat(reverse, notNullValue());
        assertThat(reverse.getItem(), is(1));
        assertThat(reverse.getNext(), nullValue());

    }

    @Test
    public void shouldReverseMoreThenOneNodeList() {
        //given
        ListNode<Integer> head = new ListNode(1);
        head.setNext(new ListNode(2));
        head.getNext().setNext(new ListNode(3));

        //when
        ListNode<Integer> reverse = reverseList.reverse(head);

        //then
        assertThat(reverse, notNullValue());
        assertThat(reverse.getItem(), is(3));
        assertThat(reverse.getNext().getItem(), is(2));
        assertThat(reverse.getNext().getNext().getItem(), is(1));
        assertThat(reverse.getNext().getNext().getNext(), nullValue());

    }
}

File: src/test/scala/com/algo/scala/linkedlist/ReverseListTest.scala
package com.algo.scala.linkedlist

import com.aglo.scala.linkedlist.{ListNode, NonEmptyLinkedList, LinkedList, EmptyLinkedList}
import org.scalatest.FunSuite


class ReverseListTest extends FunSuite {

  def emptyList: LinkedList = new EmptyLinkedList

  def singleNodeList(value: Int): LinkedList = new NonEmptyLinkedList(new ListNode(value), emptyList)

  def reverse(list: LinkedList): LinkedList = list.reverse(emptyList)

  def node(value: Int): ListNode = new ListNode(value)

  test("Reverse Empty List") {
    //given
    def empty = emptyList

    //when
    def reversed = reverse(empty)

    //then
    assert(reversed.size == 0)
  }

  test("Reverse Single Node List") {
    //given
    def singleList = singleNodeList(1)

    //when
    def reversed = reverse(singleList)

    //then
    assert(reversed.isEqual(singleNodeList(1)))
  }

  test("Reverse Two Nodes List") {
    //given
    def list = singleNodeList(2).addToHead(node(1))

    //when
    def reversed = reverse(list)

    //then
    assert(reversed.isEqual(singleNodeList(1).addToHead(node(2))))
  }

}
